src/service_discovery/mod.rs

```diff
@@ -130,26 +121,11 @@ impl ServiceDiscovery {
         }
     }
 
-    fn handle_incoming_msg(
-        &mut self,
-        bytes_rxd: usize,
-        peer_addr: SocketAddr,
-        core: &mut Core,
-        poll: &Poll,
-    ) {
-        let msg: DiscoveryMsg = match deserialise(&self.read_buf[..bytes_rxd]) {
-            Ok(msg) => msg,
-            Err(e) => {
-                debug!("Bogus message serialisation error: {:?}", e);
-                return;
-            }
-        };
-
+    fn handle_incoming_msg(&mut self, msg: DiscoveryMsg, peer_addr: SocketAddr) {
         match msg {
             DiscoveryMsg::Request { guid } => {
                 if self.listen && self.guid != guid {
-                    self.reply_to.push_back(peer_addr);
-                    self.write(core, poll)
+                    self.respond_with_their_addr(peer_addr);
```

Comment:
Why not just use `self.write()` which takes an `Option` and pass it `Some()` ? Makes it uniform with code elsewhere in all our libs. Also a single place to read from and write to a socket per state is better as we need to only look there for anything. So bring the functionality of `respond_with..()` over here inline and remove that newly created function.